### DISCUSSION TITLE
Run a self-build image with --userns=keep-id, then the image size will increase

### DISCUSSION THREAD

**Discussion by phiblar** [containers/podman#24118] [General]
# Issue Description 
An image is  generated from a Dockerfile, on the condition that no user is specified. As soon as the container is created and started with the --userns=keep-id parameter, the image size nearly doubles. What is the reason for the increase in image size? Are there alternative solutions which I can use? 
# Steps to reproduce the issue 
1.  ```bash
    $ podman build --squash --tag fedora.buildenv -f /path/to/Dockerfile
    ```
2.  ```bash
    $ podman images
    ```
3.  ```bash
    $ podman run -di --userns=keep-id localhost/fedora.buildenv:latest 
    ```
4.  ```bash 
    $ podman images
    ```

# Describe the results you received

**before podman run**
```bash
$ podman images
REPOSITORY                         TAG               IMAGE ID      CREATED        SIZE
localhost/fedora.buildenv          latest            edbe624b0ac9  3 minutes ago  9.52 GB
```

**after podman run**
```bash
$ podman images
REPOSITORY                         TAG               IMAGE ID      CREATED        SIZE
localhost/fedora.buildenv          latest            edbe624b0ac9  6 minutes ago  18.8 GB
```

**Comment by Luap99** [comment]
Because as rootless we cannot use idmapped mounts so when you create a userns all the files must be copied and then chown'ed to the new uid/gids which is also why the first run with --userns=keep-id is so slow to start up, see https://github.com/containers/podman/blob/main/docs/tutorials/performance.md#choosing-a-storage-driver

**Reply by phiblar** [reply]
Thank you. Is it possible to create the user namespace with podman build or in the Dockerfile itself, without having to copy all files and change their ownership to the new UID/GIDs as soon as the container is started with --userns=keep-id?

**Reply by jwcard** [reply]
Yes, I am also interested in this as well. I am trying to keep the size on disk as minimal as possible. I am creating the image using the same user that will run the container although I do use `--uidmap` and `--guidmap` and I tried using the **_userns-uid-map_** option to `podman build` hoping to create an image that would be run as a container with the "chown" already performed and hence only have a single copy. But that didn't have any effect on the final image size.

I also thought about using **_podman create_** to create the image and then **_podman commit_** followed by **_podman save_** but that didn't help at all.

In my case, I don't care about the startup time but do care about the disk space and unfortunately, I need to use rootless.

I'm wondering if exporting the container and then running `chown` and then importing that back would help? Is this just an unsolvable issue in a rootless mode?

**Comment by eriksjolund** [comment]
An alternative solution is to use _fuse-overlayfs_ instead of _native overlay_

I listed the image size with `podman images` after running 

```
podman run --rm --userns=keep-id registry.fedoraproject.org/fedora /bin/true
```


| type | image size after `podman run --rm --userns=keep-id ...` |
| ---- |          --- |
| native overlay | 636 Mb |
| fuse-overlayfs | 318 Mb |



<details>
  <summary>Click to expand all the details of the tests</summary>

### native overlay 

1. Create user _test1_
   ```
   sudo useradd test1
   ```
2. Open an interactive shell session for user _test1_
   ```
   sudo machinectl shell --uid=test1
   ```
3. Pull the image 
   ```
   podman pull -q registry.fedoraproject.org/fedora
   ```
4. List images
   ```
   podman images
   ```
   The following output is printed
   ```
   REPOSITORY                         TAG         IMAGE ID      CREATED       SIZE
   registry.fedoraproject.org/fedora  latest      a8accd4685fe  23 hours ago  318 MB
   ```
   __result:__ the size of the image is listed as __318 MB__
5. Run podman with `--userns=keep-id`
   ```
   podman run --rm --userns=keep-id registry.fedoraproject.org/fedora /bin/true
   ```
6. List images
   ```
   podman images
   ```
   The following output is printed
   ```
   REPOSITORY                         TAG         IMAGE ID      CREATED       SIZE
   registry.fedoraproject.org/fedora  latest      a8accd4685fe  23 hours ago  636 MB
   ```
   __result:__ the size of the image is listed as __636 MB__

### fuse-overlayfs 

1. Create user _test2_
   ```
   sudo useradd test2
   ```
2. Open an interactive shell session for user _test2_
   ```
   sudo machinectl shell --uid=test2
   ```
3. Create directory
   ```
   mkdir -p ~/.config/containers
   ```
4. Create the file _~/.config/containers/storage.conf_ containing
   ```
   [storage]
   driver = "overlay"
   [storage.options.overlay]
   mount_program = "/usr/bin/fuse-overlayfs"
   ```
5. Pull the image 
   ```
   podman pull -q registry.fedoraproject.org/fedora
   ```
6. List images
   ```
   podman images
   ```
   The following output is printed
   ```
   REPOSITORY                         TAG         IMAGE ID      CREATED       SIZE
   registry.fedoraproject.org/fedora  latest      a8accd4685fe  23 hours ago  318 MB
   ```
   __result:__ the size of the image is listed as __318 MB__
7. Run podman with `--userns=keep-id`
   ```
   podman run --rm --userns=keep-id registry.fedoraproject.org/fedora /bin/true
   ```
8. List images
   ```
   podman images
   ```
   The following output is printed
   ```
   REPOSITORY                         TAG         IMAGE ID      CREATED       SIZE
   registry.fedoraproject.org/fedora  latest      a8accd4685fe  23 hours ago  318 MB
   ```
   __result:__ the size of the image is listed as __318 MB__




</details>



__Sidenote:__

I also did a test with `--userns=auto` and _native overlay_

```
podman pull -q registry.fedoraproject.org/fedora
for i in {0..9}; do
  podman run --rm --userns=auto --pull=never -d registry.fedoraproject.org/fedora sleep inf
done
```

The image size then increased roughly 10 times.

**Comment by jwcard** [comment]
Thank you! This does solve two problems. It solves the size problem and the
time it takes to start up the first instance of the container.
I tried to make the change to /etc/containers/storage.conf but that didn't
seem to work but no big deal making the local change to the user config
indeed resolves this issue.
